### PR TITLE
Always return metadata in get/search APIs.

### LINKED ISSUE #8068: Return _parent value by default for child documents
When you search or GET to retrieve documents you don't get back the value for `_parent` field (or `_routing` for that matter) which is part of the essential metadata. Meaning the document cannot be retrieved or updated without the value. This is breaking some functionality, example use cases:

* reindex (results of `scan` being fed into `bulk`) cannot work
* persistence layers on top of elasticsearch cannot update a document they retrieved from ES without additional info/queries
* you cannot GET a document you searched for (for example to retrieve fields you excluded)

I believe it would be worth the cost to retrieve the value of `_parent` field when retrieving a child document.

### PR DESCRIPTION
**Pull request by jpountz**
This commit makes the get and search APIs always return `_parent`, `_routing`,
`_timestamp` and `_ttl` in addition to `_id` and `_type`. This way, consumers
always have all required information in order to reindex a document.

Closes #8068

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/explain/ExplainActionTests.java [hunk 1]
```diff
@@ -138,7 +139,7 @@ public void testExplainWithFields() throws Exception {
         assertThat(response.getExplanation().getValue(), equalTo(1.0f));
         assertThat(response.getGetResult().isExists(), equalTo(true));
         assertThat(response.getGetResult().getId(), equalTo("1"));
-        assertThat(response.getGetResult().getFields().size(), equalTo(1));
+        assertThat(response.getGetResult().getFields().size(), greaterThanOrEqualTo(1));
```
Comment (colings86) [inline core/src/test/java/org/elasticsearch/explain/ExplainActionTests.java]: Why is this `greaterThanOrEqualTo()`? Do we not know how many metadata fields we should get returned here?

Comment (jpountz) [inline core/src/test/java/org/elasticsearch/explain/ExplainActionTests.java]: We have random index templates that enable `_timestamp` for instance, which will be part of the list of return fields now that we force it to be returned when enabled.

Comment (colings86) [inline core/src/test/java/org/elasticsearch/explain/ExplainActionTests.java]: ok, I didn't know that, thanks. I wonder if there is some way we can get the list of fields that have been randomly added? The problem I see here is that we aren't really testing that only the requesting fields are returned.

Comment (jpountz) [inline core/src/test/java/org/elasticsearch/explain/ExplainActionTests.java]: I could change the assertion to something like `fields.keySet().containsAll(Arrays.asList(expectedField1, expectedField2))`?

**Comment by jpountz** [comment]
@colings86 I pushed a new commit

File: core/src/main/java/org/elasticsearch/index/get/ShardGetService.java [hunk 2]
```diff
@@ -201,58 +195,72 @@ private GetResult innerGet(String type, String id, String[] gFields, boolean rea
...
+                    }
+                    Object value = null;
+                    if (field.equals(RoutingFieldMapper.NAME)) {
+                        value = source.routing;
+                    } else if (field.equals(ParentFieldMapper.NAME) && docMapper.parentFieldMapper().active()) {
+                        value = source.parent;
+                    } else if (field.equals(TimestampFieldMapper.NAME) && docMapper.timestampFieldMapper().enabled()) {
+                        value = source.timestamp;
+                    } else if (field.equals(TTLFieldMapper.NAME) && docMapper.TTLFieldMapper().enabled()) {
+                        // Call value for search with timestamp + ttl here to display the live remaining ttl value and be consistent with the search result display
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: to -> so

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/get/ShardGetService.java]: Oops nevermind, I misread.

**Comment by rjernst** [comment]
LGTM

**Pull request merged**